File: VeilLink/Security/ReplayWindow.swift
import Foundation

struct ReplayWindow {
    private(set) var highestSequence: UInt64 = 0
    private var seenBitmap: UInt64 = 0

    func isPotentiallyFresh(_ sequence: UInt64) -> Bool {
        guard sequence > 0 else { return false }
        guard highestSequence > 0 else { return true }
        if sequence > highestSequence { return true }
        let distance = highestSequence - sequence
        guard distance < 64 else { return false }
        return (seenBitmap & (UInt64(1) << distance)) == 0
    }

    mutating func accept(_ sequence: UInt64) -> Bool {
        guard isPotentiallyFresh(sequence) else { return false }
        if highestSequence == 0 {
            highestSequence = sequence
            seenBitmap = 1
            return true
        }
        if sequence > highestSequence {
            let shift = sequence - highestSequence
            if shift >= 64 {
                seenBitmap = 1
            } else {
                seenBitmap = (seenBitmap << shift) | 1
            }
            highestSequence = sequence
            return true
        }
        let distance = highestSequence - sequence
        seenBitmap |= UInt64(1) << distance
        return true
    }
}

File: android/protocol-core/src/main/kotlin/studio/zeo/veillink/protocol/ReplayWindow.kt
package studio.zeo.veillink.protocol

class ReplayWindow {
    var highestSequence: ULong = 0u
        private set
    private var seenBitmap: ULong = 0u

    fun isPotentiallyFresh(sequence: ULong): Boolean {
        if (sequence == 0uL) return false
        if (highestSequence == 0uL) return true
        if (sequence > highestSequence) return true
        val distance = highestSequence - sequence
        if (distance >= 64u) return false
        return (seenBitmap and (1uL shl distance.toInt())) == 0uL
    }

    fun accept(sequence: ULong): Boolean {
        if (!isPotentiallyFresh(sequence)) return false
        if (highestSequence == 0uL) {
            highestSequence = sequence
            seenBitmap = 1u
            return true
        }
        if (sequence > highestSequence) {
            val shift = sequence - highestSequence
            seenBitmap = if (shift >= 64u) 1u else (seenBitmap shl shift.toInt()) or 1u
            highestSequence = sequence
            return true
        }
        val distance = highestSequence - sequence
        seenBitmap = seenBitmap or (1uL shl distance.toInt())
        return true
    }
}
